
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Lightweight Python Library for Bacterial Genomics">
      
      
      
        <link rel="canonical" href="https://tomdstanton.github.io/baclib/reference/baclib/engines/variant/">
      
      
        <link rel="prev" href="../sparse/">
      
      
        <link rel="next" href="../../io/">
      
      
        
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>variant - baclib ðŸ¦ ðŸ§¬ðŸ’»</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="deep-orange" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#baclib.engines.variant" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="baclib ðŸ¦ ðŸ§¬ðŸ’»" class="md-header__button md-logo" aria-label="baclib ðŸ¦ ðŸ§¬ðŸ’»" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19.82 14A1.746 1.746 0 0 0 23 13a1.746 1.746 0 0 0-3.18-1H19c0-1.57-.5-3-1.4-4.19l1.34-1.34c.56.1 1.13-.06 1.56-.47.67-.69.67-1.8 0-2.5-.69-.67-1.8-.67-2.5 0-.41.43-.57 1-.47 1.56L16.19 6.4c-.92-.69-2-1.15-3.19-1.32v-1.4A1.746 1.746 0 0 0 12 .5a1.746 1.746 0 0 0-1 3.18v1.4c-.9.13-1.74.42-2.5.86L7.39 4.35c.19-.52.14-1.12-.2-1.6a1.75 1.75 0 0 0-2.44-.43c-.79.55-.99 1.64-.43 2.43.34.49.88.75 1.43.75l1.18 1.68c-.43.43-.77.91-1.06 1.44A1.76 1.76 0 0 0 4 9c-.67.7-.67 1.8 0 2.5.29.27.64.43 1 .5 0 .54.07 1.06.18 1.56l-1.31.35c-.42-.35-1-.5-1.58-.35a1.75 1.75 0 1 0 .92 3.38c.57-.16 1-.58 1.18-1.1l1.51-.41c.45.79 1.05 1.49 1.75 2.07l-1.1 2c-.55.08-1.05.39-1.34.92a1.749 1.749 0 1 0 3.08 1.66c.28-.52.27-1.12.02-1.61l1.07-1.97c.81.32 1.69.5 2.62.5h.18c-.13.26-.18.56-.18.88.08.97.92 1.69 1.88 1.62s1.69-.92 1.62-1.88c-.04-.5-.29-.94-.65-1.23.47-.21.92-.48 1.34-.79l2.34 2.34c-.1.56.06 1.13.47 1.56.7.67 1.8.67 2.5 0s.67-1.8 0-2.5c-.43-.41-1-.57-1.56-.47l-2.34-2.34c.49-.65.87-1.39 1.11-2.19zm-9.32-2c-.83 0-1.5-.67-1.5-1.5S9.67 9 10.5 9s1.5.67 1.5 1.5-.67 1.5-1.5 1.5m3.5 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            baclib ðŸ¦ ðŸ§¬ðŸ’»
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              variant
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="deep-orange" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="deep-orange" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/tomdstanton/baclib" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/baclib
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="baclib ðŸ¦ ðŸ§¬ðŸ’»" class="md-nav__button md-logo" aria-label="baclib ðŸ¦ ðŸ§¬ðŸ’»" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19.82 14A1.746 1.746 0 0 0 23 13a1.746 1.746 0 0 0-3.18-1H19c0-1.57-.5-3-1.4-4.19l1.34-1.34c.56.1 1.13-.06 1.56-.47.67-.69.67-1.8 0-2.5-.69-.67-1.8-.67-2.5 0-.41.43-.57 1-.47 1.56L16.19 6.4c-.92-.69-2-1.15-3.19-1.32v-1.4A1.746 1.746 0 0 0 12 .5a1.746 1.746 0 0 0-1 3.18v1.4c-.9.13-1.74.42-2.5.86L7.39 4.35c.19-.52.14-1.12-.2-1.6a1.75 1.75 0 0 0-2.44-.43c-.79.55-.99 1.64-.43 2.43.34.49.88.75 1.43.75l1.18 1.68c-.43.43-.77.91-1.06 1.44A1.76 1.76 0 0 0 4 9c-.67.7-.67 1.8 0 2.5.29.27.64.43 1 .5 0 .54.07 1.06.18 1.56l-1.31.35c-.42-.35-1-.5-1.58-.35a1.75 1.75 0 1 0 .92 3.38c.57-.16 1-.58 1.18-1.1l1.51-.41c.45.79 1.05 1.49 1.75 2.07l-1.1 2c-.55.08-1.05.39-1.34.92a1.749 1.749 0 1 0 3.08 1.66c.28-.52.27-1.12.02-1.61l1.07-1.97c.81.32 1.69.5 2.62.5h.18c-.13.26-.18.56-.18.88.08.97.92 1.69 1.88 1.62s1.69-.92 1.62-1.88c-.04-.5-.29-.94-.65-1.23.47-.21.92-.48 1.34-.79l2.34 2.34c-.1.56.06 1.13.47 1.56.7.67 1.8.67 2.5 0s.67-1.8 0-2.5c-.43-.41-1-.57-1.56-.47l-2.34-2.34c.49-.65.87-1.39 1.11-2.19zm-9.32-2c-.83 0-1.5-.67-1.5-1.5S9.67 9 10.5 9s1.5.67 1.5 1.5-.67 1.5-1.5 1.5m3.5 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1"/></svg>

    </a>
    baclib ðŸ¦ ðŸ§¬ðŸ’»
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tomdstanton/baclib" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    tomdstanton/baclib
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    API Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>baclib
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>baclib
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    baclib
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1_2" >
        
          
          <label class="md-nav__link" for="__nav_2_1_2" id="__nav_2_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>apis
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>apis
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../apis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    apis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../apis/datasets/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>datasets
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../apis/prodoric/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>prodoric
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1_3" >
        
          
          <label class="md-nav__link" for="__nav_2_1_3" id="__nav_2_1_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>containers
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>containers
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    containers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/alignment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alignment
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/cluster/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>cluster
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/feature/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>feature
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/genome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>genome
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/graph/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>graph
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/motif/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>motif
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/mutations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>mutations
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/qualifier/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>qualifier
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/record/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>record
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../containers/seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>seq
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1_4" >
        
          
          <label class="md-nav__link" for="__nav_2_1_4" id="__nav_2_1_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>core
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>core
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    core
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../core/alphabet/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>alphabet
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../core/interval/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>interval
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1_5" checked>
        
          
          <label class="md-nav__link" for="__nav_2_1_5" id="__nav_2_1_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>engines
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_1_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_1_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>engines
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    engines
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../index_py/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>index
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../motif/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>motif
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pairwise/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>pairwise
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sparse/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>sparse
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>variant
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>variant
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller" class="md-nav__link">
    <span class="md-ellipsis">
      
        VariantCaller
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VariantCaller">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller.call_pileup_variants" class="md-nav__link">
    <span class="md-ellipsis">
      
        call_pileup_variants
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller.call_variants" class="md-nav__link">
    <span class="md-ellipsis">
      
        call_variants
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1_6" >
        
          
          <label class="md-nav__link" for="__nav_2_1_6" id="__nav_2_1_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>io
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/genbank/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>genbank
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/motif/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>motif
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/seq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>seq
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../io/tabular/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    <code class="doc-symbol doc-symbol-nav doc-symbol-module"></code>tabular
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller" class="md-nav__link">
    <span class="md-ellipsis">
      
        VariantCaller
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VariantCaller">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller.call_pileup_variants" class="md-nav__link">
    <span class="md-ellipsis">
      
        call_pileup_variants
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#baclib.engines.variant.VariantCaller.call_variants" class="md-nav__link">
    <span class="md-ellipsis">
      
        call_variants
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<div class="doc doc-object doc-module">



<h1 id="baclib.engines.variant" class="doc doc-heading">
            <code>baclib.engines.variant</code>


</h1>

    <div class="doc doc-contents first">

        <p>Variant calling engine for identifying SNPs, indels, and pileup-based consensus variants from alignments.</p>










<div class="doc doc-children">









<div class="doc doc-object doc-class">



<h2 id="baclib.engines.variant.VariantCaller" class="doc doc-heading">
            <code>VariantCaller</code>


</h2>


    <div class="doc doc-contents ">



        <p>Engine for calling variants from alignments.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>baclib/engines/variant.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">VariantCaller</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Engine for calling variants from alignments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">call_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">:</span> <span class="n">AlignmentBatch</span><span class="p">,</span> <span class="n">queries</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span> <span class="n">targets</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MutationBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Identifies mutations (SNPs, Indels) from an AlignmentBatch.</span>

<span class="sd">        Args:</span>
<span class="sd">            alignments: The alignment batch.</span>
<span class="sd">            queries: The query sequences (SeqBatch).</span>
<span class="sd">            targets: The target sequences (SeqBatch).</span>

<span class="sd">        Returns:</span>
<span class="sd">            A MutationBatch containing all identified variants.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># We need to iterate alignments and parse CIGARs</span>
        <span class="c1"># This is hard to fully vectorize because CIGARs are variable length strings</span>
        <span class="c1"># and produce variable numbers of mutations.</span>

        <span class="n">mut_starts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_ends</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_strands</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_refs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_alts</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Access raw arrays for speed</span>
        <span class="n">q_data</span><span class="p">,</span> <span class="n">q_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">arrays</span>
        <span class="n">t_data</span><span class="p">,</span> <span class="n">t_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">arrays</span>

        <span class="c1"># Iterate alignments</span>
        <span class="c1"># Note: We assume alignments are valid and indices match the batches</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)):</span>
            <span class="c1"># Get Alignment Metadata</span>
            <span class="n">q_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">t_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

            <span class="c1"># Global offsets in SeqBatch</span>
            <span class="n">q_offset</span> <span class="o">=</span> <span class="n">q_starts</span><span class="p">[</span><span class="n">q_idx</span><span class="p">]</span>
            <span class="n">t_offset</span> <span class="o">=</span> <span class="n">t_starts</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>

            <span class="c1"># Alignment start coords (0-based on sequence)</span>
            <span class="n">q_aln_start</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">t_aln_start</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">t_strand</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_strands</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

            <span class="n">cigar</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">cigars</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">cigar</span><span class="p">:</span> <span class="k">continue</span>

            <span class="c1"># Parse CIGAR</span>
            <span class="c1"># We track current position in Query and Target</span>
            <span class="n">curr_q</span> <span class="o">=</span> <span class="n">q_aln_start</span>
            <span class="n">curr_t</span> <span class="o">=</span> <span class="n">t_aln_start</span>

            <span class="n">ops</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">Cigar</span><span class="o">.</span><span class="n">parse_into_arrays</span><span class="p">(</span><span class="n">cigar</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ops</span><span class="p">)):</span>
                <span class="n">op</span> <span class="o">=</span> <span class="n">ops</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                <span class="n">count</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">EQ</span><span class="p">:</span> <span class="c1"># Match (=)</span>
                    <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                    <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>
                <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">X</span><span class="p">:</span> <span class="c1"># Mismatch (X)</span>
                    <span class="c1"># We have &#39;count&#39; mismatches. </span>
                    <span class="c1"># Ideally we check base-by-base to see if they are contiguous SNPs or MNP</span>
                    <span class="c1"># For simplicity, we treat contiguous X as one MNP block</span>

                    <span class="c1"># Extract Ref (Target) and Alt (Query)</span>
                    <span class="c1"># Note: We need to handle strand if query is RC. </span>
                    <span class="c1"># But usually AlignmentBatch stores coords relative to forward strand of SeqBatch entry.</span>

                    <span class="n">ref_slice</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                    <span class="n">alt_slice</span> <span class="o">=</span> <span class="n">q_data</span><span class="p">[</span><span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="p">:</span> <span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>

                    <span class="c1"># Create Mutation</span>
                    <span class="c1"># Interval is on Target (Ref)</span>
                    <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                    <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">)</span>
                    <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                    <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_slice</span><span class="p">))</span>
                    <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_slice</span><span class="p">))</span>

                    <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                    <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

                <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">I</span><span class="p">:</span> <span class="c1"># Insertion (in Query, gap in Target)</span>
                    <span class="n">alt_slice</span> <span class="o">=</span> <span class="n">q_data</span><span class="p">[</span><span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="p">:</span> <span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                    <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                    <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                    <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                    <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                    <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_slice</span><span class="p">))</span>
                    <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>

                <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">D</span><span class="p">:</span> <span class="c1"># Deletion (gap in Query, present in Target)</span>
                    <span class="n">ref_slice</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                    <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                    <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">)</span>
                    <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                    <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_slice</span><span class="p">))</span>
                    <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                    <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

                <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">M</span><span class="p">:</span> <span class="c1"># Match or Mismatch (Ambiguous)</span>
                    <span class="c1"># We need to compare sequences to find SNPs</span>
                    <span class="c1"># This is expensive in Python loop. </span>
                    <span class="c1"># Ideally we use extended CIGAR (=/X) from the start.</span>
                    <span class="c1"># If forced to use M, we would need a kernel to scan q/t arrays.</span>
                    <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                    <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

                <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">S</span><span class="p">:</span> <span class="c1"># Soft Clip</span>
                    <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                    <span class="c1"># Soft clips consume query but not target, and are not mutations</span>

        <span class="c1"># Construct Batch</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">mut_starts</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">MutationBatch</span><span class="o">.</span><span class="n">empty</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">MutationBatch</span><span class="p">(</span>
            <span class="n">IntervalBatch</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_starts</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_ends</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_strands</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_refs</span><span class="p">),</span>
            <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_alts</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">call_pileup_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">:</span> <span class="n">AlignmentBatch</span><span class="p">,</span>
                             <span class="n">queries</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span> <span class="n">targets</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span>
                             <span class="n">min_depth</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MutationBatch</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calls consensus variants (SNPs/Deletions) based on alignment pileup statistics.</span>

<span class="sd">        Args:</span>
<span class="sd">            alignments: The alignment batch.</span>
<span class="sd">            queries: The query sequences (SeqBatch).</span>
<span class="sd">            targets: The target sequences (SeqBatch).</span>
<span class="sd">            min_depth: Minimum coverage depth to call a variant.</span>
<span class="sd">            min_freq: Minimum allele frequency (count / depth) to call a variant.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A MutationBatch of consensus variants.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mut_starts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_ends</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_strands</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_refs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_alts</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">q_data</span><span class="p">,</span> <span class="n">q_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">arrays</span>
        <span class="n">t_data</span><span class="p">,</span> <span class="n">t_starts</span><span class="p">,</span> <span class="n">t_lengths</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">arrays</span>

        <span class="c1"># RC Table for handling reverse strand alignments</span>
        <span class="n">rc_table</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">complement</span>
        <span class="k">if</span> <span class="n">rc_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Fallback identity if no complement (e.g. Protein), though pileup usually implies DNA</span>
            <span class="n">rc_table</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>

        <span class="c1"># Process per target to keep memory usage low</span>
        <span class="k">for</span> <span class="n">t_idx</span><span class="p">,</span> <span class="n">group_indices</span> <span class="ow">in</span> <span class="n">alignments</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">by_target</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">t_len</span> <span class="o">=</span> <span class="n">t_lengths</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>
            <span class="n">t_offset</span> <span class="o">=</span> <span class="n">t_starts</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>

            <span class="c1"># Counts: [Pos, Base]. Base 0-3 (ACGT), 4 (Deletion)</span>
            <span class="c1"># Assuming DNA alphabet size 4. If larger, we need to adjust.</span>
            <span class="n">n_sym</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
            <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">t_len</span><span class="p">,</span> <span class="n">n_sym</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

            <span class="c1"># 1. Accumulate Counts</span>
            <span class="k">for</span> <span class="n">aln_i</span> <span class="ow">in</span> <span class="n">group_indices</span><span class="p">:</span>
                <span class="n">q_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_indices</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
                <span class="n">q_offset</span> <span class="o">=</span> <span class="n">q_starts</span><span class="p">[</span><span class="n">q_idx</span><span class="p">]</span>

                <span class="n">q_s</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_starts</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
                <span class="n">t_s</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_starts</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_strands</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span> <span class="c1"># Relative strand</span>

                <span class="n">cigar</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">cigars</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">cigar</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">_pileup_add_kernel</span><span class="p">(</span>
                    <span class="n">counts</span><span class="p">,</span> <span class="n">q_data</span><span class="p">,</span> <span class="n">q_offset</span><span class="p">,</span> <span class="n">q_s</span><span class="p">,</span> <span class="n">t_s</span><span class="p">,</span> 
                    <span class="n">cigar</span><span class="p">,</span> <span class="n">Cigar</span><span class="o">.</span><span class="n">_BYTE_TO_OP</span><span class="p">,</span>
                    <span class="n">strand</span><span class="p">,</span> <span class="n">rc_table</span>
                <span class="p">)</span>

            <span class="c1"># 2. Call Variants</span>
            <span class="c1"># We scan the counts matrix for non-ref alleles</span>
            <span class="n">ref_seq_data</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">t_len</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">t_len</span><span class="p">):</span>
                <span class="n">total</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">total</span> <span class="o">&lt;</span> <span class="n">min_depth</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">ref_base</span> <span class="o">=</span> <span class="n">ref_seq_data</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

                <span class="c1"># Check for variants</span>
                <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_sym</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">ref_base</span><span class="p">:</span> <span class="k">continue</span>

                    <span class="n">count</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">pos</span><span class="p">,</span> <span class="n">allele</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">continue</span>

                    <span class="n">freq</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="n">total</span>
                    <span class="k">if</span> <span class="n">freq</span> <span class="o">&gt;=</span> <span class="n">min_freq</span><span class="p">:</span>
                        <span class="c1"># Found a variant!</span>
                        <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
                        <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

                        <span class="c1"># Ref Seq</span>
                        <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_seq_data</span><span class="p">[</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span><span class="o">+</span><span class="mi">1</span><span class="p">]))</span>

                        <span class="c1"># Alt Seq</span>
                        <span class="k">if</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">n_sym</span><span class="p">:</span> <span class="c1"># Deletion</span>
                            <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># We need to construct a 1-byte array for the alt base</span>
                            <span class="n">alt_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">allele</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>
                            <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_arr</span><span class="p">))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">mut_starts</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">MutationBatch</span><span class="o">.</span><span class="n">empty</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">MutationBatch</span><span class="p">(</span>
            <span class="n">IntervalBatch</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_starts</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_ends</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_strands</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_refs</span><span class="p">),</span>
            <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_alts</span><span class="p">)</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="baclib.engines.variant.VariantCaller.call_pileup_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">call_pileup_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">queries</span><span class="p">,</span> <span class="n">targets</span><span class="p">,</span> <span class="n">min_depth</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">min_freq</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Calls consensus variants (SNPs/Deletions) based on alignment pileup statistics.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>alignments</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="AlignmentBatch (baclib.containers.alignment.AlignmentBatch)" href="../../containers/alignment/#baclib.containers.alignment.AlignmentBatch">AlignmentBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The alignment batch.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>queries</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="SeqBatch (baclib.containers.seq.SeqBatch)" href="../../containers/seq/#baclib.containers.seq.SeqBatch">SeqBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The query sequences (SeqBatch).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>targets</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="SeqBatch (baclib.containers.seq.SeqBatch)" href="../../containers/seq/#baclib.containers.seq.SeqBatch">SeqBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The target sequences (SeqBatch).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_depth</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum coverage depth to call a variant.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_freq</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum allele frequency (count / depth) to call a variant.</p>
              </div>
            </td>
            <td>
                  <code>0.2</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="baclib.containers.delta.MutationBatch">MutationBatch</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A MutationBatch of consensus variants.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>baclib/engines/variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">call_pileup_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">:</span> <span class="n">AlignmentBatch</span><span class="p">,</span>
                         <span class="n">queries</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span> <span class="n">targets</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span>
                         <span class="n">min_depth</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">min_freq</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MutationBatch</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calls consensus variants (SNPs/Deletions) based on alignment pileup statistics.</span>

<span class="sd">    Args:</span>
<span class="sd">        alignments: The alignment batch.</span>
<span class="sd">        queries: The query sequences (SeqBatch).</span>
<span class="sd">        targets: The target sequences (SeqBatch).</span>
<span class="sd">        min_depth: Minimum coverage depth to call a variant.</span>
<span class="sd">        min_freq: Minimum allele frequency (count / depth) to call a variant.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A MutationBatch of consensus variants.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mut_starts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_ends</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_strands</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_refs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_alts</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">q_data</span><span class="p">,</span> <span class="n">q_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">arrays</span>
    <span class="n">t_data</span><span class="p">,</span> <span class="n">t_starts</span><span class="p">,</span> <span class="n">t_lengths</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">arrays</span>

    <span class="c1"># RC Table for handling reverse strand alignments</span>
    <span class="n">rc_table</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">complement</span>
    <span class="k">if</span> <span class="n">rc_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Fallback identity if no complement (e.g. Protein), though pileup usually implies DNA</span>
        <span class="n">rc_table</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>

    <span class="c1"># Process per target to keep memory usage low</span>
    <span class="k">for</span> <span class="n">t_idx</span><span class="p">,</span> <span class="n">group_indices</span> <span class="ow">in</span> <span class="n">alignments</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">by_target</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">t_len</span> <span class="o">=</span> <span class="n">t_lengths</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>
        <span class="n">t_offset</span> <span class="o">=</span> <span class="n">t_starts</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>

        <span class="c1"># Counts: [Pos, Base]. Base 0-3 (ACGT), 4 (Deletion)</span>
        <span class="c1"># Assuming DNA alphabet size 4. If larger, we need to adjust.</span>
        <span class="n">n_sym</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">t_len</span><span class="p">,</span> <span class="n">n_sym</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

        <span class="c1"># 1. Accumulate Counts</span>
        <span class="k">for</span> <span class="n">aln_i</span> <span class="ow">in</span> <span class="n">group_indices</span><span class="p">:</span>
            <span class="n">q_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_indices</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
            <span class="n">q_offset</span> <span class="o">=</span> <span class="n">q_starts</span><span class="p">[</span><span class="n">q_idx</span><span class="p">]</span>

            <span class="n">q_s</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_starts</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
            <span class="n">t_s</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_starts</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_strands</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span> <span class="c1"># Relative strand</span>

            <span class="n">cigar</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">cigars</span><span class="p">[</span><span class="n">aln_i</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">cigar</span><span class="p">:</span> <span class="k">continue</span>

            <span class="n">_pileup_add_kernel</span><span class="p">(</span>
                <span class="n">counts</span><span class="p">,</span> <span class="n">q_data</span><span class="p">,</span> <span class="n">q_offset</span><span class="p">,</span> <span class="n">q_s</span><span class="p">,</span> <span class="n">t_s</span><span class="p">,</span> 
                <span class="n">cigar</span><span class="p">,</span> <span class="n">Cigar</span><span class="o">.</span><span class="n">_BYTE_TO_OP</span><span class="p">,</span>
                <span class="n">strand</span><span class="p">,</span> <span class="n">rc_table</span>
            <span class="p">)</span>

        <span class="c1"># 2. Call Variants</span>
        <span class="c1"># We scan the counts matrix for non-ref alleles</span>
        <span class="n">ref_seq_data</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">t_len</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">t_len</span><span class="p">):</span>
            <span class="n">total</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">total</span> <span class="o">&lt;</span> <span class="n">min_depth</span><span class="p">:</span> <span class="k">continue</span>

            <span class="n">ref_base</span> <span class="o">=</span> <span class="n">ref_seq_data</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>

            <span class="c1"># Check for variants</span>
            <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_sym</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">ref_base</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">count</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">pos</span><span class="p">,</span> <span class="n">allele</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="k">continue</span>

                <span class="n">freq</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="n">total</span>
                <span class="k">if</span> <span class="n">freq</span> <span class="o">&gt;=</span> <span class="n">min_freq</span><span class="p">:</span>
                    <span class="c1"># Found a variant!</span>
                    <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
                    <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

                    <span class="c1"># Ref Seq</span>
                    <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_seq_data</span><span class="p">[</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span><span class="o">+</span><span class="mi">1</span><span class="p">]))</span>

                    <span class="c1"># Alt Seq</span>
                    <span class="k">if</span> <span class="n">allele</span> <span class="o">==</span> <span class="n">n_sym</span><span class="p">:</span> <span class="c1"># Deletion</span>
                        <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># We need to construct a 1-byte array for the alt base</span>
                        <span class="n">alt_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">allele</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>
                        <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_arr</span><span class="p">))</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">mut_starts</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">MutationBatch</span><span class="o">.</span><span class="n">empty</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">MutationBatch</span><span class="p">(</span>
        <span class="n">IntervalBatch</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_starts</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_ends</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_strands</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_refs</span><span class="p">),</span>
        <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_alts</span><span class="p">)</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="baclib.engines.variant.VariantCaller.call_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">call_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">queries</span><span class="p">,</span> <span class="n">targets</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Identifies mutations (SNPs, Indels) from an AlignmentBatch.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>alignments</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="AlignmentBatch (baclib.containers.alignment.AlignmentBatch)" href="../../containers/alignment/#baclib.containers.alignment.AlignmentBatch">AlignmentBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The alignment batch.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>queries</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="SeqBatch (baclib.containers.seq.SeqBatch)" href="../../containers/seq/#baclib.containers.seq.SeqBatch">SeqBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The query sequences (SeqBatch).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>targets</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="SeqBatch (baclib.containers.seq.SeqBatch)" href="../../containers/seq/#baclib.containers.seq.SeqBatch">SeqBatch</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The target sequences (SeqBatch).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="baclib.containers.delta.MutationBatch">MutationBatch</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A MutationBatch containing all identified variants.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>baclib/engines/variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">call_variants</span><span class="p">(</span><span class="n">alignments</span><span class="p">:</span> <span class="n">AlignmentBatch</span><span class="p">,</span> <span class="n">queries</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">,</span> <span class="n">targets</span><span class="p">:</span> <span class="n">SeqBatch</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">MutationBatch</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identifies mutations (SNPs, Indels) from an AlignmentBatch.</span>

<span class="sd">    Args:</span>
<span class="sd">        alignments: The alignment batch.</span>
<span class="sd">        queries: The query sequences (SeqBatch).</span>
<span class="sd">        targets: The target sequences (SeqBatch).</span>

<span class="sd">    Returns:</span>
<span class="sd">        A MutationBatch containing all identified variants.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># We need to iterate alignments and parse CIGARs</span>
    <span class="c1"># This is hard to fully vectorize because CIGARs are variable length strings</span>
    <span class="c1"># and produce variable numbers of mutations.</span>

    <span class="n">mut_starts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_ends</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_strands</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_refs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mut_alts</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Access raw arrays for speed</span>
    <span class="n">q_data</span><span class="p">,</span> <span class="n">q_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">queries</span><span class="o">.</span><span class="n">arrays</span>
    <span class="n">t_data</span><span class="p">,</span> <span class="n">t_starts</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">arrays</span>

    <span class="c1"># Iterate alignments</span>
    <span class="c1"># Note: We assume alignments are valid and indices match the batches</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)):</span>
        <span class="c1"># Get Alignment Metadata</span>
        <span class="n">q_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">t_idx</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="c1"># Global offsets in SeqBatch</span>
        <span class="n">q_offset</span> <span class="o">=</span> <span class="n">q_starts</span><span class="p">[</span><span class="n">q_idx</span><span class="p">]</span>
        <span class="n">t_offset</span> <span class="o">=</span> <span class="n">t_starts</span><span class="p">[</span><span class="n">t_idx</span><span class="p">]</span>

        <span class="c1"># Alignment start coords (0-based on sequence)</span>
        <span class="n">q_aln_start</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">q_starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">t_aln_start</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_starts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">t_strand</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">t_strands</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="n">cigar</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">cigars</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">cigar</span><span class="p">:</span> <span class="k">continue</span>

        <span class="c1"># Parse CIGAR</span>
        <span class="c1"># We track current position in Query and Target</span>
        <span class="n">curr_q</span> <span class="o">=</span> <span class="n">q_aln_start</span>
        <span class="n">curr_t</span> <span class="o">=</span> <span class="n">t_aln_start</span>

        <span class="n">ops</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">Cigar</span><span class="o">.</span><span class="n">parse_into_arrays</span><span class="p">(</span><span class="n">cigar</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ops</span><span class="p">)):</span>
            <span class="n">op</span> <span class="o">=</span> <span class="n">ops</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="n">count</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">EQ</span><span class="p">:</span> <span class="c1"># Match (=)</span>
                <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>
            <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">X</span><span class="p">:</span> <span class="c1"># Mismatch (X)</span>
                <span class="c1"># We have &#39;count&#39; mismatches. </span>
                <span class="c1"># Ideally we check base-by-base to see if they are contiguous SNPs or MNP</span>
                <span class="c1"># For simplicity, we treat contiguous X as one MNP block</span>

                <span class="c1"># Extract Ref (Target) and Alt (Query)</span>
                <span class="c1"># Note: We need to handle strand if query is RC. </span>
                <span class="c1"># But usually AlignmentBatch stores coords relative to forward strand of SeqBatch entry.</span>

                <span class="n">ref_slice</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                <span class="n">alt_slice</span> <span class="o">=</span> <span class="n">q_data</span><span class="p">[</span><span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="p">:</span> <span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>

                <span class="c1"># Create Mutation</span>
                <span class="c1"># Interval is on Target (Ref)</span>
                <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">)</span>
                <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_slice</span><span class="p">))</span>
                <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_slice</span><span class="p">))</span>

                <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">I</span><span class="p">:</span> <span class="c1"># Insertion (in Query, gap in Target)</span>
                <span class="n">alt_slice</span> <span class="o">=</span> <span class="n">q_data</span><span class="p">[</span><span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="p">:</span> <span class="n">q_offset</span> <span class="o">+</span> <span class="n">curr_q</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">alt_slice</span><span class="p">))</span>
                <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">D</span><span class="p">:</span> <span class="c1"># Deletion (gap in Query, present in Target)</span>
                <span class="n">ref_slice</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="p">:</span> <span class="n">t_offset</span> <span class="o">+</span> <span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">]</span>
                <span class="n">mut_starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span><span class="p">)</span>
                <span class="n">mut_ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_t</span> <span class="o">+</span> <span class="n">count</span><span class="p">)</span>
                <span class="n">mut_strands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t_strand</span><span class="p">)</span>
                <span class="n">mut_refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">seq_from</span><span class="p">(</span><span class="n">ref_slice</span><span class="p">))</span>
                <span class="n">mut_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">empty_seq</span><span class="p">())</span>
                <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">M</span><span class="p">:</span> <span class="c1"># Match or Mismatch (Ambiguous)</span>
                <span class="c1"># We need to compare sequences to find SNPs</span>
                <span class="c1"># This is expensive in Python loop. </span>
                <span class="c1"># Ideally we use extended CIGAR (=/X) from the start.</span>
                <span class="c1"># If forced to use M, we would need a kernel to scan q/t arrays.</span>
                <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                <span class="n">curr_t</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">op</span> <span class="o">==</span> <span class="n">CigarOp</span><span class="o">.</span><span class="n">S</span><span class="p">:</span> <span class="c1"># Soft Clip</span>
                <span class="n">curr_q</span> <span class="o">+=</span> <span class="n">count</span>
                <span class="c1"># Soft clips consume query but not target, and are not mutations</span>

    <span class="c1"># Construct Batch</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">mut_starts</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">MutationBatch</span><span class="o">.</span><span class="n">empty</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">MutationBatch</span><span class="p">(</span>
        <span class="n">IntervalBatch</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_starts</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_ends</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
            <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mut_strands</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">targets</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_refs</span><span class="p">),</span>
        <span class="n">queries</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">batch_from</span><span class="p">(</span><span class="n">mut_alts</span><span class="p">)</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>







  
  






                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../../..", "features": ["toc.follow", "content.action.edit"], "search": "../../../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>